src/core/xref.js

```diff
@@ -0,0 +1,119 @@
+);
+
+/**
+ * main external reference driver
+ * @param {Object} conf respecConfig
+ * @param {Array:Elements} elems possibleExternalLinks
+ */
+export async function run(conf, elems) {
+  const { xref } = conf;
+  const xrefMap = createXrefMap(elems);
+
+  const query = createXrefQuery(xrefMap);
+  const apiURL =
+    (xref && typeof xref === "object" && new URL(xref.url)) || API_URL;
+  const results = await fetchXrefs(query, apiURL);
+
+  addDataCiteToTerms(results, xrefMap, conf);
+}
+
+/**
+ * maps term to elements and context
+ * @param {Array:Elements} elems
+ * @returns {Map} term => [ { elem } ]
+ */
+function createXrefMap(elems) {
+  return elems.reduce((map, elem) => {
+    const term = normalize(elem.textContent);
+    const xrefsForTerm = map.has(term) ? map.get(term) : [];
+    xrefsForTerm.push({ elem });
+    return map.set(term, xrefsForTerm);
+  }, new Map());
+}
+
+/**
+ * creates a body for POST request to API
+ * @param {Map} xrefs
+ * @returns {Object} { keys: [{ term }] }
+ */
+function createXrefQuery(xrefs) {
+  const queryKeys = [...xrefs.keys()].reduce((queryKeys, term) => {
+    return queryKeys.add(JSON.stringify({ term }));
+  }, new Set());
+  return { keys: [...queryKeys].map(JSON.parse) };
+}
+
+// fetch from network
+async function fetchXrefs(query, url) {
+  const options = {
+    method: url.host === location.host ? "GET" : "POST",
+    body: JSON.stringify(query),
+    headers: {
+      "Content-Type": "application/json",
+    },
+  };
+  const json = await fetch(url, options);
+  return await json.json();
+}
+
+/**
+ * adds data-cite attributes to elems
+ * for each term from conf.xref[term] for which results are found.
+ * @param {Object} results parsed JSON results returned from API
+ * @param {Map} xrefMap xrefMap
+ * @param {Object} conf respecConfig
+ */
+function addDataCiteToTerms(results, xrefMap, conf) {
+  for (const [term, entries] of xrefMap) {
+    entries.forEach(entry => {
+      const { elem } = entry;
+
+      const result = disambiguate(results[term], entry, term);
+      if (!result) return;
+
+      const { uri, spec: cite, normative } = result;
+      const path = uri.includes("/") ? uri.split("/", 1)[1] : uri;
+      const [citePath, citeFrag] = path.split("#");
+      Object.assign(elem.dataset, { cite, citePath, citeFrag });
+
+      if (normative == true) conf.normativeReferences.add(cite);
+    });
+  }
+}
+
+// disambiguate fetched results based on context
+function disambiguate(data, context, term) {
+  const { elem } = context;
+
+  if (!data || !data.length) {
+    const msg = `No external reference data found for term \`${term}\`.`;
+    showError(msg, elem);
+    return null;
+  }
+
+  if (data.length === 1) {
+    return data[0]; // unambiguous
+  }
+
+  const msg = `Ambiguity in data found for term \`${term}\`.`;
+  showError(msg, elem);
+  return null;
+
+  function showError(msg, elem) {
```

Comment:
nit: `showInlineError()` ... we should maybe move this to utils, as we do this a lot in the code base.